Failed to execute goal `com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor`

# Defect
Dear Stefan Ferstl (@ferstl),

I have just tried to upgrade the Maven plugin for my current Maven multi-module project.

After performing the Maven plugin upgrade, the `depgraph:reactor` goal execution (`mvn depgraph:reactor`) fails with the following error message:

```
[ERROR] Failed to execute goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor (default-cli) on project custom-maven-project: Execution default-cli of goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor failed: Cannot invoke "String.equals(Object)" because the return value of "com.github.ferstl.depgraph.dependency.puml.PumlNodeInfo.getStereotype()" is null -> [Help 1]
```

## Maven plugin configuration
```xml
<project>
    <build>
        <pluginManagement>
            <plugins>
                <plugin>
                    <groupId>com.github.ferstl</groupId>
                    <artifactId>depgraph-maven-plugin</artifactId>
                    <version>4.0.0</version>
                    <configuration>
                        <graphFormat>puml</graphFormat>
                        <scope>compile</scope>
                        <excludes>
                            <exclude>org.slf4j:slf4j-api</exclude>
                            <exclude>com.google.guava:guava</exclude>
                            <exclude>io.vavr:vavr</exclude>
                        </excludes>
                        <transitiveExcludes>
                            <transitiveExclude>*</transitiveExclude>
                        </transitiveExcludes>
                    </configuration>
                </plugin>
            </plugins>
        </pluginManagement>
    </build>
</project>
```

## Detailed error message (`mvn -X depgraph:reactor`)
<details>
    <summary>Detailed error message</summary>

```
[ERROR] Failed to execute goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor (default-cli) on project custom-maven-project: Execution default-cli of goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor failed: Cannot invoke "String.equals(Object)" because the return value of "com.github.ferstl.depgraph.dependency.puml.PumlNodeInfo.getStereotype()" is null -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor (default-cli) on project custom-maven-project: Execution default-cli of goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor failed: Cannot invoke "String.equals(Object)" because the return value of "com.github.ferstl.depgraph.dependency.puml.PumlNodeInfo.getStereotype()" is null
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:78)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal com.github.ferstl:depgraph-maven-plugin:4.0.0:reactor failed: Cannot invoke "String.equals(Object)" because the return value of "com.github.ferstl.depgraph.dependency.puml.PumlNodeInfo.getStereotype()" is null
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:78)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException: Cannot invoke "String.equals(Object)" because the return value of "com.github.ferstl.depgraph.dependency.puml.PumlNodeInfo.getStereotype()" is null
    at com.github.ferstl.depgraph.graph.puml.PumlGraphFormatter.writeNodes (PumlGraphFormatter.java:71)
    at com.github.ferstl.depgraph.graph.puml.PumlGraphFormatter.format (PumlGraphFormatter.java:38)
    at com.github.ferstl.depgraph.graph.GraphBuilder.toString (GraphBuilder.java:149)
    at com.github.ferstl.depgraph.dependency.ReactorGraphFactory.createGraph (ReactorGraphFactory.java:45)
    at com.github.ferstl.depgraph.AbstractGraphMojo.execute (AbstractGraphMojo.java:209)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:78)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

</details>

Best regards,
Sergey Vyacheslavovich Brunov.

--- a/src/main/java/com/github/ferstl/depgraph/dependency/puml/PumlDependencyEgdeRenderer.java
+++ b/src/main/java/com/github/ferstl/depgraph/dependency/puml/PumlDependencyEgdeRenderer.java
@@ -18,7 +18,6 @@ package com.github.ferstl.depgraph.dependency.puml;
 import com.github.ferstl.depgraph.dependency.DependencyNode;
 import com.github.ferstl.depgraph.dependency.NodeResolution;
 import com.github.ferstl.depgraph.graph.EdgeRenderer;
-
 import static com.github.ferstl.depgraph.dependency.VersionAbbreviator.abbreviateVersion;
 
 /**
@@ -44,7 +43,7 @@ public class PumlDependencyEgdeRenderer implements EdgeRenderer<DependencyNode> 
 
     switch (resolution) {
       case INCLUDED:
-
+      case PARENT:
         edgeInfo.withBegin("-[")
             .withColor(INCLUDE_COLOR)
             .withEnd("]->")

--- a/src/main/java/com/github/ferstl/depgraph/graph/puml/PumlGraphFormatter.java
+++ b/src/main/java/com/github/ferstl/depgraph/graph/puml/PumlGraphFormatter.java
@@ -68,7 +68,7 @@ public class PumlGraphFormatter implements GraphFormatter {
           .append(escape(node.getNodeId()));
 
 
-      if (!nodeInfo.getStereotype().equals(SCOPE_COMPILE)) {
+      if (nodeInfo.getStereotype() != null && !nodeInfo.getStereotype().equals(SCOPE_COMPILE)) {
         puml.append("<<")
             .append(nodeInfo.getStereotype())
             .append(">>");

--- a/src/test/java/com/github/ferstl/depgraph/ReactorIntegrationTest.java
+++ b/src/test/java/com/github/ferstl/depgraph/ReactorIntegrationTest.java
@@ -49,4 +49,44 @@ public class ReactorIntegrationTest {
 
     assertFileContents(basedir, "expectations/reactor.txt", "target/dependency-graph.txt");
   }
+
+  @Test
+  public void reactorDot() throws Exception {
+    File basedir = this.resources.getBasedir("reactor-graph");
+    MavenExecutionResult result = this.mavenRuntime
+        .forProject(basedir)
+        .withCliOption("-DgraphFormat=dot")
+        .execute("clean", "depgraph:reactor");
+
+    result.assertErrorFreeLog();
+
+    assertFileContents(basedir, "expectations/reactor.dot", "target/dependency-graph.dot");
+  }
+
+  // Issue #144
+  @Test
+  public void reactorPuml() throws Exception {
+    File basedir = this.resources.getBasedir("reactor-graph");
+    MavenExecutionResult result = this.mavenRuntime
+        .forProject(basedir)
+        .withCliOption("-DgraphFormat=puml")
+        .execute("clean", "depgraph:reactor");
+
+    result.assertErrorFreeLog();
+
+    assertFileContents(basedir, "expectations/reactor.puml", "target/dependency-graph.puml");
+  }
+
+  @Test
+  public void reactorGml() throws Exception {
+    File basedir = this.resources.getBasedir("reactor-graph");
+    MavenExecutionResult result = this.mavenRuntime
+        .forProject(basedir)
+        .withCliOption("-DgraphFormat=gml")
+        .execute("clean", "depgraph:reactor");
+
+    result.assertErrorFreeLog();
+
+    assertFileContents(basedir, "expectations/reactor.gml", "target/dependency-graph.gml");
+  }
 }

--- a/src/test/projects/reactor-graph/expectations/reactor.dot
+++ b/src/test/projects/reactor-graph/expectations/reactor.dot
@@ -1,0 +1,21 @@
+digraph "G" {
+  node [shape="box",style="rounded",fontname="Helvetica",fontsize="14"]
+  edge [fontsize="10",fontname="Helvetica"]
+
+  // Node Definitions:
+  "com.github.ferstl:reactor-ui"[label=<reactor-ui>]
+  "com.github.ferstl:reactor-application"[label=<reactor-application>]
+  "com.github.ferstl:reactor-database"[label=<reactor-database>]
+  "com.github.ferstl:reactor-service"[label=<reactor-service>]
+  "com.github.ferstl:reactor-common"[label=<reactor-common>]
+  "com.github.ferstl:reactor-parent"[label=<reactor-parent>]
+  "com.github.ferstl:reactor-api"[label=<reactor-api>]
+
+  // Edge Definitions:
+  "com.github.ferstl:reactor-ui" -> "com.github.ferstl:reactor-application"[style="solid"]
+  "com.github.ferstl:reactor-database" -> "com.github.ferstl:reactor-service"[style="solid"]
+  "com.github.ferstl:reactor-common" -> "com.github.ferstl:reactor-database"[style="solid"]
+  "com.github.ferstl:reactor-common" -> "com.github.ferstl:reactor-ui"[style="solid"]
+  "com.github.ferstl:reactor-parent" -> "com.github.ferstl:reactor-api"[style="solid"]
+  "com.github.ferstl:reactor-parent" -> "com.github.ferstl:reactor-common"[style="solid"]
+}

--- a/src/test/projects/reactor-graph/expectations/reactor.gml
+++ b/src/test/projects/reactor-graph/expectations/reactor.gml
@@ -1,0 +1,67 @@
+graph [
+node [
+id "com.github.ferstl:reactor-ui"
+label "reactor-ui"
+]
+
+node [
+id "com.github.ferstl:reactor-application"
+label "reactor-application"
+]
+
+node [
+id "com.github.ferstl:reactor-database"
+label "reactor-database"
+]
+
+node [
+id "com.github.ferstl:reactor-service"
+label "reactor-service"
+]
+
+node [
+id "com.github.ferstl:reactor-common"
+label "reactor-common"
+]
+
+node [
+id "com.github.ferstl:reactor-parent"
+label "reactor-parent"
+]
+
+node [
+id "com.github.ferstl:reactor-api"
+label "reactor-api"
+]
+
+edge [
+source "com.github.ferstl:reactor-ui"
+target "com.github.ferstl:reactor-application"
+]
+
+edge [
+source "com.github.ferstl:reactor-database"
+target "com.github.ferstl:reactor-service"
+]
+
+edge [
+source "com.github.ferstl:reactor-common"
+target "com.github.ferstl:reactor-database"
+]
+
+edge [
+source "com.github.ferstl:reactor-common"
+target "com.github.ferstl:reactor-ui"
+]
+
+edge [
+source "com.github.ferstl:reactor-parent"
+target "com.github.ferstl:reactor-api"
+]
+
+edge [
+source "com.github.ferstl:reactor-parent"
+target "com.github.ferstl:reactor-common"
+]
+
+]

--- a/src/test/projects/reactor-graph/expectations/reactor.puml
+++ b/src/test/projects/reactor-graph/expectations/reactor.puml
@@ -1,0 +1,22 @@
+@startuml
+skinparam defaultTextAlignment center
+skinparam rectangle {
+  BackgroundColor<<optional>> beige
+  BackgroundColor<<test>> lightGreen
+  BackgroundColor<<runtime>> lightBlue
+  BackgroundColor<<provided>> lightGray
+}
+rectangle "reactor-ui" as com_github_ferstl_reactor_ui
+rectangle "reactor-application" as com_github_ferstl_reactor_application
+rectangle "reactor-database" as com_github_ferstl_reactor_database
+rectangle "reactor-service" as com_github_ferstl_reactor_service
+rectangle "reactor-common" as com_github_ferstl_reactor_common
+rectangle "reactor-parent" as com_github_ferstl_reactor_parent
+rectangle "reactor-api" as com_github_ferstl_reactor_api
+com_github_ferstl_reactor_ui -[#000000]-> com_github_ferstl_reactor_application
+com_github_ferstl_reactor_database -[#000000]-> com_github_ferstl_reactor_service
+com_github_ferstl_reactor_common -[#000000]-> com_github_ferstl_reactor_database
+com_github_ferstl_reactor_common -[#000000]-> com_github_ferstl_reactor_ui
+com_github_ferstl_reactor_parent -[#000000]-> com_github_ferstl_reactor_api
+com_github_ferstl_reactor_parent -[#000000]-> com_github_ferstl_reactor_common
+@enduml

